fcli-core/fcli-ssc/src/main/java/com/fortify/cli/ssc/appversion/cli/cmd/SSCAppVersionCopyStateCommand.java

```diff
@@ -0,0 +1,92 @@
+import com.fortify.cli.common.output.cli.cmd.IJsonNodeSupplier;
+import com.fortify.cli.common.output.cli.mixin.OutputHelperMixins;
+import com.fortify.cli.common.output.transform.IActionCommandResultSupplier;
+import com.fortify.cli.common.output.transform.IRecordTransformer;
+import com.fortify.cli.common.rest.unirest.UnexpectedHttpResponseException;
+import com.fortify.cli.ssc._common.output.cli.cmd.AbstractSSCJsonNodeOutputCommand;
+import com.fortify.cli.ssc._common.rest.SSCUrls;
+import com.fortify.cli.ssc.appversion.cli.mixin.SSCAppAndVersionNameResolverMixin;
+import com.fortify.cli.ssc.appversion.cli.mixin.SSCCopyFromAppVersionResolverMixin;
+import com.fortify.cli.ssc.appversion.helper.SSCAppVersionDescriptor;
+import com.fortify.cli.ssc.appversion.helper.SSCAppVersionHelper;
+import kong.unirest.UnirestInstance;
+import lombok.Getter;
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Mixin;
+
+import java.util.HashMap;
+
+@Command(name = "copy-issues")
+public class SSCAppVersionCopyStateCommand extends AbstractSSCJsonNodeOutputCommand implements IJsonNodeSupplier, IRecordTransformer, IActionCommandResultSupplier {
+    @Getter
+    @Mixin
+    private OutputHelperMixins.TableNoQuery outputHelper;
+    @Mixin
+    private SSCAppAndVersionNameResolverMixin.PositionalParameter sscAppAndVersionNameResolver;
+    @Mixin
+    private SSCCopyFromAppVersionResolverMixin.RequiredOption fromAppVersionResolver;
+
+    @Override
+    public JsonNode getJsonNode(UnirestInstance unirest) {
+        ObjectMapper mapper = new ObjectMapper();
+        SSCAppVersionDescriptor targetAppVersionDescriptor = SSCAppVersionHelper.getRequiredAppVersion(unirest, sscAppAndVersionNameResolver.getAppAndVersionName(), sscAppAndVersionNameResolver.getDelimiter());
+        SSCAppVersionDescriptor sourceAppVersionDescriptor = fromAppVersionResolver.getAppVersionDescriptor(unirest, sscAppAndVersionNameResolver.getDelimiter());
+
+        return mapper.valueToTree(copyState(unirest, sourceAppVersionDescriptor, targetAppVersionDescriptor));
+
+    }
+
+    @Override
+    public JsonNode transformRecord(JsonNode record) {
+        return SSCAppVersionHelper.renameFields(record);
+    }
+
+    @Override
+    public String getActionCommandResult() {
+        return "COPY_REQUESTED";
+    }
+
+    @Override
+    public boolean isSingular() {
+        return true;
+    }
+
+    private static final HashMap<String, String> copyState(UnirestInstance unirest, SSCAppVersionDescriptor sourceAppVersion, SSCAppVersionDescriptor targetAppVersion) {
+        HashMap<String, String> appVersionIds = new HashMap<String, String>();
+        appVersionIds.put("previousProjectVersionId", sourceAppVersion.getVersionId());
+        appVersionIds.put("projectVersionId", targetAppVersion.getVersionId());
+        String action = "COPY_REQUESTED";
+        try {
+            JsonNode response = unirest.post(SSCUrls.PROJECT_VERSIONS_ACTION_COPY_CURRENT_STATE)
```

Comment:
I tried on 22.2.3, 23.1.1 and 23.1.2, and the current url works fine for both :
- ssc av create app:it1 --copy-from app2:it2 --copy Issues --skip-if-exists
- ssc av copy-issues app:ver --copy-from app2:ver

Reply:
Did you try copying to an appversion that already has scan results available? Maybe SSC actively rejects this (also given my comments above)?

Reply:
I just did with 22.2 and 23.1, and with totally different fpr

Reply:
I just tried as well, indeed didn't encounter any issues. @alemhg, as you mentioned an error when trying to copy state, can you enable trace logging on fcli and share either the full log with us, or just the HTTP response code and body returned by SSC on the /projectVersions/action/copyCurrentState REST call to see why you were having issues with this operation?

Reply:
Hello @rsenden @Andhrei 

After doing the call projectVersions/action/copyCurrentState" sending body "{ \"previousProjectVersionId\": 74473564, \"projectVersionId\": 179757774}" the responseCode is 200 and the vulnerabilities are copied.

Using FCLI after session login:

- when I try to create a new application version and copy the issues from another existing application version (ssc av create app:it1 --copy-from app2:it2 --copy Issues --skip-if-exists), the new application version is created but the issues (vulnerabilities) are not copied (see attached log create-command-log.txt)
- when I try to copy the issues from one application version to another one (ssc av copy-issues app:ver --copy-from app2:ver) the copy fails (ACTION=COPY_FAILED) (See attached log copy-issues-command-log.txt).

Please let me know if you need anything else.

Best regards,
Alejo

[copy-issues-command-log.txt](https://github.com/fortify/fcli/files/13361710/copy-issues-command-log.txt)
[create-command-log.txt](https://github.com/fortify/fcli/files/13361711/create-command-log.txt)

Reply:
@alemhg Thanks for sharing the additional details. I've downloaded the logs and we'll have a look as to why this might be failing. I've updated your post to remove the links to the logs, as these may contain private/sensitive data like host names and tokens.

Reply:
@alemhg I've had a look at your log files. All requests to the `copyFromPartial` and `copyCurrentState` endpoints (and also `authEntities` endpoint) are failing with a `401 Authentication failed` error. 

Apparently, with the default SSC configuration these endpoints are only accessible using a `UnifiedLoginToken`, which means that in fcli you'd need to authenticate with username/password or with a pre-generated `UnifiedLoginToken` (which is not really viable given the maximum lifetime of 1 day).

SSC 23.2 will add a new `AutomationToken`, which has similar permissions as `UnifiedLoginToken` but with a maximum lifetime of 365 days. Given the security implications, you'd want to use dedicated user accounts with proper permissions managed through user roles to generate such tokens; you wouldn't want to generate an `AutomationToken` using an Administrator account and then use this token for pipeline integrations.

I think you have a self-hosted SSC instance, in which case there is an option to modify the default token types & permissions in `<tomcat root>/webapps/ssc/WEB-INF/internal/serviceContext.xml`. You could either add these endpoints to an existing token definition (like `CIToken`), or add an `AutomationToken` definition similar to the upcoming SSC 23.2; for future compatibility the latter is probably best.

For reference, following is the `AutomationToken` definition in SSC 23.2:

```xml
	<bean id="automationToken" class="com.fortify.manager.security.ws.AuthenticationTokenSpec">
		<property name="key" value="AutomationToken"/>
		<property name="capabilityDescription">
			<value>
				<![CDATA[
				This token specification provides the capability to access most of the REST API endpoints permitted to its issuing user.
				Intended for use with longer running automations.
				WARNING: Due to the access this token provides and its maximum allowed lifetime, you must take extra care securing the token to reduce risk of API misuse or unintended use.
				It is highly recommended that you evaluate the planned use of this token and ensure you limit the life of the token based on your environments' tolerance for risk.
				]]>
			</value>
		</property>
		<property name="maxDaysToLive" value="365" />
		<property name="actionPermitted">
			<list value-type="java.lang.String">
				<value>(GET|PUT|POST|DELETE|HEAD|OPTIONS)=/api/v\d+/.*</value>
				<value>POST=/d3srv</value>
			</list>
		</property>
		<property name="terminalActions">
			<list value-type="java.lang.String">
			</list>
		</property>
	</bean>
```

Reply:
Is this "bug" going to get fixed in 23.2? There is no reason why copy_current_state should not be run by a CI_TOKEN? 

As mentioned by @alemhg above, this endpoint does work fine with the CI_TOKEN and is used by the Fortify SSC UI, that's how we found it:
"/api/v1/projectVersions/" + str(projectVersionId) + "/action"
{ "type": "COPY_CURRENT_STATE", "values": { "projectVersionId": projectVersionId, "previousProjectVersionId": previousProjectVersionId } }

Reply:
Hi @fcoene, good to hear from you again, it's been a long time. Indeed, the generic action endpoint seems to be supported for CITokens, so probably better to use this in fcli then. I'm just waiting for confirmation from SSC engineering that both approaches offer the exact same functionality, also with regards to copying 'issue introduced date' in SSC 23.2.

Looking at the CIToken definition in `serviceContext.xml` for SSC 23.2, it contains this entry which I guess fixes this 'bug':
`POST=/api/v\d+/projectVersions/(\d+/)?action/[a-zA-Z]+/?`

But since we try to keep fcli compatible with all recent SSC versions, it's probably still better to use the generic action endpoint if it provides the exact same functionality, to allow these fcli commands/options to be used with CIToken on older SSC versions.

Looks like the `PUT /api/v1/projectVersions/<id>/authEntities` endpoint is not supported for CIToken in either SSC 23.1 or 23.2, so application version user assignment options in fcli will likely still fail silently when using a CIToken.